### ISSUE TITLE
adding extensions for multiple annotations (admin only)

### ISSUE THREAD

**Issue by Antonialock** [pombase/canto#1131]
currently, to add several independent extensions to the same annotation you need to either a) use the "edit as string"  or b) create multiple annotations.

but a) is slow and b) is time-consuming and horrible

so, I want a way to add multiple pipe-separated extensions using the nice extensions UI
this should be admin only to avoid confusing people

I want this at fairly hogh priority because it is a bit annoying

**Comment by kimrutherford** [comment]
What should the interface look like for this?

**Comment by Antonialock** [comment]
Hmm, as a start, can you make the extension editable? At the moment you need to delete the extension + add a new one in order to change it. 

In the example below for instance, say I added the extension "telomere" but meant to add "rDNA". In order to change it I must delete "telomere" and then add "rDNA". It would be nicer to be able to click on "telomere" and then get to the "start typing a description" box.
The ability to edit the extension might make copy+edit less painful.

I still think it would be nice to add multiple extensions without using copy+edit. Maybe a button to "add another extension" would expand out a second form for another annotation?


![screen shot 2016-06-09 at 10 28 53](https://cloud.githubusercontent.com/assets/12727506/15924999/26060658-2e2d-11e6-8e79-d172bee5a3cc.png)

**Comment by ValWood** [comment]
+1

**Comment by ValWood** [comment]
Re:
I still think it would be nice to add multiple extensions without using copy+edit. Maybe a button to "add another extension" would expand out a second form for another annotation?


Isn't there a ticket about this as an admin only feature somewhere?

**Comment by mah11** [comment]
this *is* the ticket about this as an admin-only feature ...

**Comment by ValWood** [comment]
Doh

**Comment by kimrutherford** [comment]
> Hmm, as a start, can you make the extension editable?

I've made a separate issue for that: #1154

Assuming that was implemented, how should the rest of the interface work for multiple pipe-separated extensions?

**Comment by Antonialock** [comment]
Would it be possible to have a button that that expands out a second form to add another extension?

**Comment by kimrutherford** [comment]
Let's chat about this on Skype.

**Comment by kimrutherford** [comment]
> Would it be possible to have a button that that expands out a second form to add another extension?

Might get a bit wide for multiple extensions?

**Comment by Antonialock** [comment]
Can't you expand it vertically? 
see attached image for example. 

In this example the form has been expanded out once, and two extensions have been added to the term, effectively this will create two separate annotations. The newly expanded form would be blank initially - the two extensions are only identical in this example because I just copied the image above.

**Comment by Antonialock** [comment]
![slide1](https://cloud.githubusercontent.com/assets/12727506/16295590/b6465220-391f-11e6-9c1b-0e93c0ba7bd0.jpg)

**Comment by kimrutherford** [comment]
I've had a first try at this.  In the test Canto there is now an "Add another group ..." link when you are adding extensions.  It only shows up if you add a relation.

After clicking the link the display doesn't change much, it just shows " | [empty]" add the end of the text version of the extension (see screenshot 2 below).  If you then add another relation it will be part of a new independent extension.  It should be clear(-ish) if you give it a go.

Please suggest better wording for "Add another group ...".  Maybe "Add an independent extension"?
"[empty]" isn't great either.

There are some big limitations: you can't edit a relation (only delete it) and you can only add relations to most recent "group".  You can edit with "Edit as text" though.

There's also a bug: the relations that can't be added twice aren't being greyed out.  I'm working on that.

It's not intuitive at all but it's a start.

![screenshot 2016-07-07 18 42 20](https://cloud.githubusercontent.com/assets/90474/16644643/ac17f200-4472-11e6-84a0-3b60887c55d5.png)
![screenshot 2016-07-07 18 46 24](https://cloud.githubusercontent.com/assets/90474/16644727/2e8643f4-4473-11e6-9b41-10233a73fec4.png)

**Comment by kimrutherford** [comment]
If forgot to say: sorry Antonia, this isn't what you asked for yet - it's a work in progress.

**Comment by Antonialock** [comment]
good for a start though :-) If we allow community users to do this then maybe we want to change the pipe separation to something more obvious, at least at the point of making the annotation (seprate rows)

Add an independent extension is better than add another group but it still might not be clear...

**Comment by ValWood** [comment]
Looks good. I think in the short to medium term this will be admin only. It's a lot for users to take on board right now since everything is pretty new to them.... so its probably fine as it is. We can refine this if we decide to make it available more broadly.

**Comment by kimrutherford** [comment]
> Add an independent extension is better than add another group but it still might not be clear...

That sounds a little better so I've made that change.

**Comment by kimrutherford** [comment]
Should I release this change as is into the main Canto?

**Comment by Antonialock** [comment]
1. I can't see the option to "add an independent extension" in the test session 73bcd7e4ba3c8a8b
![screen shot 2016-07-13 at 07 57 45](https://cloud.githubusercontent.com/assets/12727506/16794515/136fb954-48d0-11e6-975f-55199c546421.png)

2. It is now possible to add several extensions of the same type in test sessions eg see screenshot

**Comment by kimrutherford** [comment]
> I can't see the option to "add an independent extension" in the test session 73bcd7e4ba3c8a8b

It worked for me.  Were you logged in as admin?

![screenshot 2016-07-13 19 10 33](https://cloud.githubusercontent.com/assets/90474/16794739/b84ec14c-492d-11e6-8a00-c62d54145245.png)

**Comment by kimrutherford** [comment]
> It is now possible to add several extensions of the same type in test sessions eg see screenshot

Sorry, I forgot I hadn't fixed that yet.  I'll do that after #1180 is done.

**Comment by Antonialock** [comment]
> Were you logged in as admin?

no :-) 

I think you can push this out for us, would be nice to have. I would like to also be able to edit individual sets of extensions but I can make a new ticket for that.

**Comment by kimrutherford** [comment]
> It is now possible to add several extensions of the same type in test sessions

That's fixed now.  I've released these changes into the main Canto.  Please let me know if you spot any weirdness with extensions.

**Comment by kimrutherford** [comment]
I'll close this now, but please open separate issues for anything else you think of.
